File: ios/Mobile Client/Mobile Client/pages/InfoProductPage.swift
//
//  InfoProductPage.swift
//  Mobile Client
//
//  Created by Eric Ruck on 4/3/18.
//  Copyright © 2018 AuditPro. All rights reserved.
//

import UIKit


/**
 * Displays the product extended information.
 * @author Eric Ruck
 */
class InfoProductPage : BasePage {
	@IBOutlet var productNameLabel: UILabel?
	@IBOutlet var brandNameLabel: UILabel?
	@IBOutlet var msrpLabel: UILabel?
	@IBOutlet var averageLabel: UILabel?
	@IBOutlet var minLabel: UILabel?
	@IBOutlet var maxLabel: UILabel?
	@IBOutlet var weighedLabel: UILabel?
	@IBOutlet var upcLabel: UILabel?
	@IBOutlet var skuLabel: UILabel?
	@IBOutlet var reorderCodeLabel: UILabel?
	@IBOutlet var categoryLabel: UILabel?
	@IBOutlet var subcategoryLabel: UILabel?
	@IBOutlet var productTypeLabel: UILabel?

	private var product: Product? = nil

	/**
	 * Initializes for the display of the passed product.
	 * @param product Product to display
	 */
	convenience init(_ product: Product) {
		self.init()
		self.product = product
	}

	/**
	 * Updates the view contents on load.
	 */
	override func viewDidLoad() {
		super.viewDidLoad()
		// Populate with passed product
		productNameLabel?.text = product?.productName
		populate(label: brandNameLabel, text: product?.brandName, altText: product?.brandNameShort)
		populate(label: msrpLabel, dollarValue: product?.msrp)
		populate(label: averageLabel, dollarValue: product?.retailPriceAverage)
		populate(label: minLabel, dollarValue: product?.retailPriceMin)
		populate(label: maxLabel, dollarValue: product?.retailPriceMax)
		populate(label: weighedLabel, formatValue: product?.randomWeight)
		populate(label: upcLabel, text: product?.upc)
		populate(label: skuLabel, text: product?.brandSku)
		populate(label: reorderCodeLabel, text: product?.currentReorderCode)
		populate(label: categoryLabel, text: product?.categoryName)
		populate(label: subcategoryLabel, text: product?.subcategoryName)
		populate(label: productTypeLabel, text: product?.productTypeName)
	}

	/**
	 * Gets the page display name.
	 * @return Page display name
	 */
	override var pageName: String? {
		return "Product Info"
	}

	/**
	 * Populates the label with a product detail.
	 * @param label Label to populate
	 * @param text Desired detail text
	 * @param altText Alternate detail text or null if none
	 */
	private func populate(label: UILabel?, text: String?, altText: String? = nil) {
		if (!(text?.trimmingCharacters(in: .whitespaces).isEmpty ?? true)) {
			label?.text = text
		} else if (!(altText?.trimmingCharacters(in: .whitespaces).isEmpty ?? true)) {
			label?.text = altText
		} else {
			label?.text = "(Not Specified)"
		}
	}

	/**
	 * Populates the label with a boolean value.
	 * @param label Label to populate
	 * @param formatValue Value to format
	 */
	private func populate(label: UILabel?, formatValue: Bool?) {
		label?.text = (formatValue == nil) ? "(NS)" : (formatValue! ? "Yes" : "No")
	}

	/**
	 * Populates the label with a product detail in dollars.
	 * @param label Label to populate
	 * @param dollarValue Dollars amount to display
	 */
	private func populate(label: UILabel?, dollarValue: Double?) {
		if (dollarValue == nil) {
			label?.text = "(NS)"
		} else {
			let formatter = NumberFormatter()
			formatter.locale = Locale.current
			formatter.numberStyle = .currency
			label?.text = formatter.string(from: NSNumber(floatLiteral: dollarValue!))
		}
	}
}

File: android/mobileclient/app/src/main/java/com/auditpro/mobile_client/pages/InfoProductPage.java
/*
  AuditPRO Mobile Client Android
  Copyright 2018 AuditPRO All Rights Reserved
 */
package com.auditpro.mobile_client.pages;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.auditpro.mobile_client.entities.Product;
import com.auditpro.mobile_client.test.R;

import java.text.NumberFormat;


/**
 * Displays the product extended information.
 * @author Eric Ruck
 */
public class InfoProductPage extends BasePage {

	/**
	 * Required empty public constructor.
	 */
	public InfoProductPage() { }

	/**
	 * Creates a new page to display the passed product.
	 * @param product Product to display
	 * @return A new instance of fragment InfoProductPage.
	 */
	public static InfoProductPage newInstance(Product product) {
		InfoProductPage fragment = new InfoProductPage();
		Bundle args = new Bundle();
		args.putParcelable(ARG_PRODUCT, product);
		fragment.setArguments(args);
		return fragment;
	}

	/**
	 * Handles creation by unpacking our arguments.
	 * @param savedInstanceState Optional state
	 */
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		if (getArguments() != null) {
			product = getArguments().getParcelable(ARG_PRODUCT);
		}
	}

	/**
	 * Creates our view.
	 * @param inflater Use to hydrate our layout
	 * @param container Parent view for page
	 * @param savedInstanceState Optional state
	 * @return Created view
	 */
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
							 Bundle savedInstanceState) {
		// Inflate the layout for this fragment
		return inflater.inflate(R.layout.fragment_info_product_page, container, false);
	}

	/**
	 * Initializes our view once created.
	 * @param view View to initialize
	 * @param savedInstanceState Optional state
	 */
	@Override
	public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);

		// Populate with passed product
		((TextView) view.findViewById(R.id.productNameText)).setText(product.getProductName());
		populate(view, R.id.brandNameText, product.getBrandName(), product.getBrandNameShort());
		populate(view, R.id.msrpText, product.getMSRP());
		populate(view, R.id.averageText, product.getRetailPriceAverage());
		populate(view, R.id.minText, product.getRetailPriceMin());
		populate(view, R.id.maxText, product.getRetailPriceMax());
		populate(view, R.id.weighedText, product.isRandomWeight());
		populate(view, R.id.upcText, product.getUPC(), null);
		populate(view, R.id.skuText, product.getBrandSKU(), null);
		populate(view, R.id.reorderCodeText, product.getCurrentReorderCode(), null);
		populate(view, R.id.categoryText, product.getCategoryName(), null);
		populate(view, R.id.subcategoryText, product.getSubcategoryName(), null);
		populate(view, R.id.productTypeText, product.getProductTypeName(), null);
	}

	/**
	 * Gets the title to display for this page.
	 * @param context Application context for string resource lookup
	 * @return Product info page title
	 */
	@Override
	public String getPageName(Context context) {
		return context.getString(R.string.message_info_title);
	}

	/**
	 * Populates the label with a product detail.
	 * @param view Parent view
	 * @param labelId Id of label to populate
	 * @param text Desired detail text
	 * @param altText Alternate detail text or null if none
	 */
	private void populate(View view, int labelId, String text, String altText) {
		TextView label = view.findViewById(labelId);
		if ((text != null) && !text.matches("^\\s*$")) {
			label.setText(text);
		} else if ((altText != null) && !altText.matches("^\\s*$")) {
			label.setText(altText);
		} else {
			label.setText(R.string.message_info_not_specified);
		}
	}

	/**
	 * Populates the label with a boolean value.
	 * @param view Parent view
	 * @param labelId Id of label to populate
	 * @param formatValue Value to format
	 */
	private void populate(View view, int labelId, boolean formatValue) {
		TextView label = view.findViewById(labelId);
		label.setText(formatValue ? R.string.button_yes : R.string.button_no);
	}

	/**
	 * Populates the label with a product detail in dollars.
	 * @param view Parent view
	 * @param labelId Id of label to populate
	 * @param dollarValue Dollars amount to display
	 */
	private void populate(View view, int labelId, Double dollarValue) {
		TextView label = view.findViewById(labelId);
		if (dollarValue == null) {
			label.setText(R.string.message_info_not_spec_short);
		} else {
			NumberFormat formatter = NumberFormat.getCurrencyInstance();
			label.setText(formatter.format(dollarValue));
		}
	}

	private static final String ARG_PRODUCT = "productArg";
	private Product product;
}
